src/components/common/FormFields.tsx

```diff
@@ -1,28 +1,154 @@
+  VisibilityOff as VisibilityOffIcon,
+} from '@material-ui/icons';
 import { useField } from 'formik';
 
-type FormikTextFieldProps = TextFieldProps & { name: string }; // make name a required prop since formik needs it
-const FormikTextField: React.FC<FormikTextFieldProps> = props => {
-  const { required, label } = props;
+import useStyles from './_jss/FormFields.jss';
+
+type FormikTextFieldProps = Omit<
+  TextFieldProps,
+  'onChange' | 'onBlur' | 'value' | 'error' | 'helperText'
+> & {
+  name: string; // make name a required prop since formik needs it
+  reserveHelperTextSpace?: boolean;
+};
+
+const FormikTextField: React.FC<FormikTextFieldProps> = React.forwardRef((props, ref) => {
+  const { required, label, reserveHelperTextSpace = true, ...additionalProps } = props;
+  const classes = useStyles();
   const [field, meta] = useField(props.name);
   const formattedLabel: ReactNode | undefined =
     required && typeof label === 'string' ? `${label}*` : label;
+  const defaultHelperText = reserveHelperTextSpace ? ' ' : undefined;
 
   return (
     <TextField
+      ref={ref}
       variant="outlined"
-      {...props}
+      {...additionalProps}
+      className={clsx(classes.formikTextField, props.className)}
+      FormHelperTextProps={{
+        classes: {
+          ...props.FormHelperTextProps?.classes,
+          root: classes.formikTextFieldHelper,
+        },
+      }}
       required={false}
       label={formattedLabel}
       onChange={field.onChange}
       onBlur={field.onBlur}
       value={field.value}
       error={meta.touched && !!meta.error}
-      helperText={meta.error}
+      helperText={(meta.touched && meta.error) || defaultHelperText}
+    />
+  );
+});
+
+type FormikPasswordProps = Omit<FormikTextFieldProps, 'inputRef' | 'InputProps' | 'type'> & {
+  /**
+   * Determines whether to display a password strength meter under the field
+   */
+  strengthMeter?: boolean;
+  /**
+   * Determines whether to display a visibility toggle switch in the field
+   */
+  visibilitySwitch?: boolean;
+};
+
+const FormikPasswordField: React.FC<FormikPasswordProps> = props => {
+  const { strengthMeter, visibilitySwitch, ...textFieldProps } = props;
+  const { name } = textFieldProps;
+  const [visible, setVisible] = useState<boolean>(false);
+  const fieldRef = useRef<HTMLDivElement | null>(null);
+  const strengthBarId = useRef<string>(uniqid('password-strength-'));
+  const [field] = useField(name);
+  const classes = useStyles();
+
+  const toggleVisibility = useCallback(() => {
+    setVisible(prevVisible => !prevVisible);
+  }, []);
+
+  useLayoutEffect(() => {
+    if (strengthMeter && fieldRef.current) {
+      const score: ZXCVBNScore = zxcvbn(field.value).score;
+      const strengthBarEl = document.getElementById(strengthBarId.current);
+      const strengthBarClasses = clsx(
+        classes.strengthBar,
+        !!field.value && (classes as Record<string, string>)[`strengthBar${score}`],
+      );
```

Comment:
instead of widening the type of `classes`, let's narrow the type of `strengthBar${score}`:
```suggestion
      const strengthBarClasses = clsx(
        classes.strengthBar,
        !!field.value &&
          classes[
            `strengthBar${score}` as
              | 'strengthBar0'
              | 'strengthBar1'
              | 'strengthBar2'
              | 'strengthBar3'
              | 'strengthBar4'
          ],
      );
```

Reply:
d5b5a8aa3ae16bd1b4b5dc96dabca75220398c12